src/libsyntax/codemap.rs

```diff
@@ -807,6 +807,70 @@ impl CodeMap {
                         hi.col.to_usize() + 1)).to_string()
     }
 
+    pub fn span_to_expanded_string(&self, sp: Span) -> String {
+        let src_str = self.span_to_string(sp);
+        let src_snip = self.span_to_snippet(sp).unwrap_or(
+            "Snippet unavailable (span malformed or source missing)".to_string());
+        if sp.expn_id == NO_EXPANSION || sp.expn_id == COMMAND_LINE_EXPN {
+            return format!("{}\n{}", src_snip, src_str);
+        }
+
+        let mut strings = vec![src_str.into_boxed_str(), src_snip.into_boxed_str()];
+        let mut cur_span = sp;
+        while cur_span.expn_id != NO_EXPANSION && cur_span.expn_id != COMMAND_LINE_EXPN {
+            let old_span = cur_span;
+            let new_span = self.with_expn_info(old_span.expn_id,
+                                              |ei| ei.map(|ei| ei.call_site));
+
+            //No info attached to src span, return
+            if let None = new_span {
+                return format!("{}\n{}", &*strings[1], &*strings[0]);
+            }
+            let new_span = new_span.unwrap();
+
+            //This compacts recursive macros down into a single entry,
+            //By only pushing sequential spans with different callees
+            let old_fmt = self.with_expn_info(
+                old_span.expn_id,
+                |ei| ei.map(|ei| ei.callee.format.clone()));
+            let new_fmt = self.with_expn_info(
+                new_span.expn_id,
+                |ei| ei.map(|ei| ei.callee.format.clone()));
+            if old_fmt == new_fmt {
+                cur_span = new_span;
+                continue;
+            }
+
+            cur_span = new_span;
+            let new_str = self.span_to_string(cur_span);
+            let new_snip = self.span_to_snippet(cur_span).unwrap_or(
+                "Snippet unavailable (span malformed or source missing)".to_string());
+            strings.push(new_str.into_boxed_str());
+            strings.push(new_snip.into_boxed_str());
+        }
+
+        //Reverse the vector (so output is in order: root -> source),
+        //and format so each span is further indented than the last
+        strings.reverse();
+        let mut indent = "".to_string();
+        let mut is_snip = true;
+        let mut output = "".to_string();
+        let mut first = true;
+        for s in strings {
+            if !first {
+                output.push_str("\n");
+            }
+            first = false;
+            output.push_str(&indent);
+            output.push_str(&*s);
+            if !is_snip {
+                indent.push_str("  ");
+            }
+            is_snip = !is_snip;
```

Comment:
I would keep two vecs - one of coords, one of snippets. You ought to be able to do this all with a bunch of iterators:

```
strs.iter().zip(snips.iter()).rev().map(|(s, snip)| format!("{}  {}", s, snip)).collect().join("\n")
```
```